src/lib/utils/hooks/SynapseAPI/trash/useTrashCan.ts

```diff
@@ -0,0 +1,117 @@
+import {
+  QueryClient,
+  useInfiniteQuery,
+  UseInfiniteQueryOptions,
+  useMutation,
+  UseMutationOptions,
+  useQueryClient,
+} from 'react-query'
+import { SynapseClient } from '../../..'
+import { SynapseClientError } from '../../../SynapseClientError'
+import { useSynapseContext } from '../../../SynapseContext'
+import { PaginatedResults, TrashedEntity } from '../../../synapseTypes'
+import {
+  invalidateAllQueriesForEntity,
+  entityQueryKeys,
+} from '../entity/queryKeys'
+
+export function useGetItemsInTrashCanInfinite(
+  options?: UseInfiniteQueryOptions<
+    PaginatedResults<TrashedEntity>,
+    SynapseClientError
+  >,
+) {
+  const { accessToken } = useSynapseContext()
+
+  return useInfiniteQuery<PaginatedResults<TrashedEntity>, SynapseClientError>(
+    ['trashcan', 'list', accessToken],
+    context => {
+      return SynapseClient.getItemsInTrashCan(accessToken, context.pageParam)
+    },
+    {
+      ...options,
+      getNextPageParam: (page, pages) => {
+        const numberOfFetchedResults = pages.flatMap(
+          page => page.results,
+        ).length
+        if (page.totalNumberOfResults! > numberOfFetchedResults) {
+          return numberOfFetchedResults
+        } else {
+          return undefined
+        }
+      },
+    },
+  )
+}
+
+export function useRestoreEntities(
+  options?: UseMutationOptions<
+    PromiseSettledResult<void>[],
+    SynapseClientError,
+    string | Set<string>
+  >,
+) {
+  const queryClient = useQueryClient()
+  const { accessToken } = useSynapseContext()
+
+  return useMutation<
+    PromiseSettledResult<void>[],
+    SynapseClientError,
+    string | Set<string>
+  >(
+    (ids: string | Set<string>) => {
+      if (typeof ids === 'string') {
+        ids = new Set([ids])
+      }
+      const promises = Array.from(ids).map(id => {
+        return SynapseClient.restoreFromTrashCan(id, accessToken)
+      })
+      return Promise.allSettled(promises)
```

Comment:
Promise.allSettled returns an array of [PromiseSettledResult](https://github.com/microsoft/TypeScript/blob/8d0c72d5e03366a7c3ed710736dbde3abed5bf63/src/lib/es2020.promise.d.ts#L11).

This lets us handle each failure individually. In contrast, Promise.all will short-circuit and reject if one promise rejects.